### ISSUE TITLE
Switching Grid selection mode from Single to Multi may break manual deselection 

### ISSUE THREAD

**Issue by kshashov** [vaadin/flow-components#3141]
### Description

Hi!

We found an issue with switching Grid selection mode from Single to Multi. In some cases it becomes impossible to deselect items manualy using checkboxes.

![grid_switch_selection_mode](https://user-images.githubusercontent.com/3830719/167271408-25b6dd63-4669-4de9-834a-78a648e9649b.gif)

The initial invistigation showed that the client-side collection `grid.SelectedItems == [null]` after the `deselectAll` invocation. And after we enable Multi selection mode and select an item again it becomes [null, <item>] which leads to further errors.

We are not exactly sure where this should be fixed: in the client component or in the flow. If necessary, we can duplicate this task in the vaadin/web-components repository.

### Expected outcome

User should be able to deselect items using selection columns.

### Minimal reproducible example

```java
        Grid<String> grid = new Grid<>();
        grid.addColumn(s -> s);
        grid.setAllRowsVisible(true);
        grid.setItems( Arrays.asList("Item 1", "Item 2", "Item 3"));

        Button deselect = new Button("Deselect All");
        deselect.addClickListener(click -> grid.deselectAll());

        Button switchSelectionMode = new Button("Multi selection mode");
        switchSelectionMode.addClickListener(click -> grid.setSelectionMode(Grid.SelectionMode.MULTI));

        add(grid, deselect, switchSelectionMode);
```

### Steps to reproduce

1. Select item in single Selection mode
2. Deselect items with `getSelectionModel().deselectAll()` 
3. Switch selection mode to 
4. Select item
5. Try to deselect item manually
5.1 Actual: item is not selected
5.2 Expected: item is selected and some errors occured in the browser console

### Environment

Vaadin version(s): 14.8.9
OS: Win10

### Browsers

Issue is not browser related

**Comment by sissbruecker** [comment]
Likely the same issue as in #324, which was fixed as part of #3008. The fix has not been back-ported to v14.

**Comment by kshashov** [comment]
Yes, it's the same issue. This task is a duplicate
